src/runtime/typemanager.cs

```diff
@@ -436,11 +456,19 @@ internal static IntPtr CreateMetaType(Type impl, out SlotsHolder slotsHolder)
             // the standard type slots, and has to subclass PyType_Type for
             // certain functions in the C runtime to work correctly with it.
 
-            IntPtr type = AllocateTypeObject("CLR Metatype");
             IntPtr py_type = Runtime.PyTypeType;
+            var heapTypeSize = (int)Marshal.ReadIntPtr(py_type, TypeOffset.tp_basicsize);
+            Debug.Assert(heapTypeSize == TypeOffset.members);
+            int metaSize = heapTypeSize + Marshal.SizeOf(typeof(ClrMetaTypeEx));
 
-            Marshal.WriteIntPtr(type, TypeOffset.tp_base, py_type);
+            IntPtr type = Runtime._PyObject_GC_Calloc(new IntPtr(metaSize));
```

Comment:
Specifically, can we use `PyType_FromSpec` here?

Reply:
Maybe not, `PyType_FromSpec` still use `PyType_Type` to allocate memory, which means it would occur the same problem as before.

Reply:
Actually, I think I solved this one. Please see https://github.com/pythonnet/pythonnet/commit/9682dc6e3b3f9231d7a801ae3202a65ed68fbcca

Reply:
No, it didn't, see https://github.com/python/cpython/blob/fbc7723778be01b8f3bb72d2dcac15ab9fbb9923/Objects/typeobject.c#L1051-L1082

Reply:
Yes, it did, see https://github.com/python/cpython/blob/8b3414818f5289eac530bf38bcfbd7b2b851805c/Include/cpython/objimpl.h#L21 :-P

Reply:
It's not about the type size which the MetaType can generate. It's about the size of MetaType itself. You can't create the MetaType by using `PyType_GenericAlloc(PyTypeType, 0)`, unless you
```c
old_size = PyTypeType.tp_basicsize;
PyTypeType.tp_basicsize += x;
PyType_GenericAlloc(PyTypeType, 0);
PyTypeType.tp_basicsize = old_size;
``` 
As long as it claimed itself as a managed type, it should have a slot indicate the place of its instances where the related GCHandle would be stored. As for now, the MetaType doesn't attach any C# object, so it remains 0. Look back to the master, if you put the PyCLRMetaType to `GetManagedObject`, actually it would read the memory it should not read(the part of the reserved `tp_itemsize`). The code doesn't explicitly call something like that, but something may trigger it does, like domain reload may visit it through PyVisit. Even it won't crash right now, but it's hard to say that's a kind of sanity code. 
Of course, you can just remove the Manged flag from PyCLRMetaType to avoid that. But I suggest following the old rule: the type of a CLR type is a managed type.
Another approach is to allocate with PyObject_Malloc, but that means it has to be initialized manually, e.g. call `_Py_NewReference` manually on debug version, add it to GCTrack.
Nevertheless, the point is, the type must clarify itself what's itself and what it can do, that commit avoided some problems, but didn't solve completely.

Reply:
As far as I can see, CLR Metatype does not contain `GCHandle`, so it does not need `GetManagedObject` to work on its instances. So I think my implementation linked above is correct. Perhaps `PyCLRMetaType` should not be a managed object then?

Reply:
It indeed has a bit of struggle with that, the PyCLRMetaType it's the ancestor of all managed types, if it claimed it's not a managed type, it against the commonsense. Currently, it doesn't need to be bound with a clr instance even it can. Well, just an identity problem.

Reply:
I think it makes sense, that CLR metatype is **not** a managed type. In my mind managed type == type, whose implementation is managed == has a managed object, that implements functionality == has `GCHandle` pointing to that object. CLR metatype does not have a `GCHandle`.